### ISSUE TITLE
fish ignores terminal newline in prompt unless it is followed by escape sequences

### ISSUE THREAD

**Issue by gnachman** [fish-shell/fish-shell#2589]
This prompt, which drops the newline after the `x`:
````
function fish_prompt
    echo x
end
````

should look the same as this prompt:
````
function fish_prompt
    echo x
    printf '\033[0m'
end
````

But the second one respects the newline after the `x` and leaves the cursor on the next line. I suspect this is not intentional—if you mean to strip trailing newlines in a prompt, they should be stripped even when followed by non-printing stuff.

**Comment by ridiculousfish** [comment]
Prompts don't have special newline handling. Instead they are run as command substitutions, which always eat a trailing newline (I'm pretty sure they do in other shells too). We can see this via count: `count (fish_prompt)` will be 1 for the first prompt, and 2 for the second.

@gnachman for my information is this causing a problem, or is it something you just noticed?

**Comment by gnachman** [comment]
It was motivated by this issue: https://gitlab.com/gnachman/iterm2/issues/4014#note_2799604
(disregard the title; the actual problem is revealed in the linked-to comment and my response)

tl;dr: A user inadvertently wrote a fish_prompt that ended with a newline. My script that wraps fish_prompt and appends an escape sequence caused his prompt to grow a newline.

Since it sounds like this is intended behavior, I should fix my code to strip the newline.

**Comment by ridiculousfish** [comment]
Stripping a newline from the output of fish_prompt before appending the escape sequence should be OK. I'd consider this behavior as intended so I'm going to close. Thanks!
